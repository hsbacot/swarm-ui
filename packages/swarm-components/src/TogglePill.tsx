import * as React from 'react';

interface Props {
	/**
	 * Indicates whether the toggle is selected
	 */
	checked?: boolean,
	/**
	 * Indicates whether the toggle is disabled
	 */
	disabled?: boolean,
	/**
	 * Text label of the content
	 */
	label?: string,
	/**
	 * Grow select to 100% width (full width)
	 */
	grow?: boolean,
	/**
	 * Text label of the content, applied if label is not set
	 */
	reset?: boolean,
	/**
	 * Strips all styles from the toggle pill
	 */
	children?: React.ReactElement,
};

const getPillState = (reset: boolean, checked: boolean) => {
	if (reset) return 'reset';
	else if (checked) return 'checked';
	else return 'unchecked';
};

const TogglePill = (props: Props) => {
	const { checked = false, disabled, label, children, grow, reset = false, ...other } = props;

	const width = grow ? 'grow' : 'default';

	return (
		<button
			data-swarm-toggle-pill={getPillState(reset, checked)}
			data-swarm-width={width}
			role="checkbox"
			aria-checked={checked}
			disabled={disabled}
			{...other}
		>
			<span>{label || children}</span>
		</button>
	);
};

export { TogglePill };
