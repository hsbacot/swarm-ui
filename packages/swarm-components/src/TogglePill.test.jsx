import React from 'react';
import { shallow } from 'enzyme';
import { TogglePill } from './TogglePill';

describe('TogglePill', () => {
	const testCases = [
		['Default', <TogglePill key="default">Toggle me</TogglePill>],
		[
			'Checked',
			<TogglePill key="checked" checked>
				Un-toggle me
			</TogglePill>,
		],
		[
			'Disabled',
			<TogglePill key="disabled" disabled>
				Can&apos;t toggle me
			</TogglePill>,
		],
		[
			'Disabled Checked',
			<TogglePill key="disabled-checked" disabled checked>
				Can&apos;t un-toggle me
			</TogglePill>,
		],
		[
			'Grow',
			<TogglePill key="grow" grow>
				Full width (grow)
			</TogglePill>,
		],
		[
			'Reset',
			<TogglePill key="reset" reset>
				No styles
			</TogglePill>,
		],
	];

	test.each(testCases)('Snapshot: %s', (description, element) => {
		expect(shallow(element)).toMatchSnapshot();
	});
});
