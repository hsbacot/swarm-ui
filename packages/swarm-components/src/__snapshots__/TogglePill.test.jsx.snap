// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TogglePill Snapshot: Checked 1`] = `
<button
  aria-checked={true}
  data-swarm-toggle-pill="checked"
  data-swarm-width="default"
  role="checkbox"
>
  <span>
    Un-toggle me
  </span>
</button>
`;

exports[`TogglePill Snapshot: Default 1`] = `
<button
  aria-checked={false}
  data-swarm-toggle-pill="unchecked"
  data-swarm-width="default"
  role="checkbox"
>
  <span>
    Toggle me
  </span>
</button>
`;

exports[`TogglePill Snapshot: Disabled 1`] = `
<button
  aria-checked={false}
  data-swarm-toggle-pill="unchecked"
  data-swarm-width="default"
  disabled={true}
  role="checkbox"
>
  <span>
    Can't toggle me
  </span>
</button>
`;

exports[`TogglePill Snapshot: Disabled Checked 1`] = `
<button
  aria-checked={true}
  data-swarm-toggle-pill="checked"
  data-swarm-width="default"
  disabled={true}
  role="checkbox"
>
  <span>
    Can't un-toggle me
  </span>
</button>
`;

exports[`TogglePill Snapshot: Grow 1`] = `
<button
  aria-checked={false}
  data-swarm-toggle-pill="unchecked"
  data-swarm-width="grow"
  role="checkbox"
>
  <span>
    Full width (grow)
  </span>
</button>
`;

exports[`TogglePill Snapshot: Reset 1`] = `
<button
  aria-checked={false}
  data-swarm-toggle-pill="reset"
  data-swarm-width="default"
  role="checkbox"
>
  <span>
    No styles
  </span>
</button>
`;
